File: BOJ/9012.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

bool solution(string input);
int main(){
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

    int T; // 입력 데이터의 수
    cin >> T;
    // 하나의 괄호 문자열의 길이는 2 이상 50 이하
    for (int tc=0; tc<T; tc++){
        string input; // '(())())'
        cin >> input;

        // solution() 호출 짝이 맞는지 검사하기
        if(solution(input)) cout << "YES" << '\n';
        else cout << "NO" << '\n';
    }

    return 0;
}

bool solution(string input){
    stack<char> tmp; // 괄호 짝이 맞는지 확인할 때 사용할 stack

    for (int i=0; i<input.length(); i++){
        if (input[i] == '('){
            tmp.push(input[i]);
        }
        else {
            if (!tmp.empty()){
                tmp.pop();
            }
            else{
                // 닫는 괄호인데 스택이 비어있으면
                // ')(())', '()())()' 이런 식인 것 -> false
                return false;
            }
        }
    }
    
    // tmp가 비어있다면 짝이 맞는것 true
    // tmp에 pop되지 못한 요소가 있다면 짝이 맞지 않는 것 false
    return tmp.empty();
}

File: BOJ/9012.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int tc = Integer.parseInt(br.readLine());
        for (int t = 0; t < tc; t++) {
            if(checkParenthesis(br.readLine())){
                sb.append("YES").append('\n');
            } else{
                sb.append("NO").append('\n');
            }
        }
        br.close();

        System.out.println(sb);
    }

    // 모든 문자열을 확인했는데, stack에 남은 원소가 있는 경우
    // stack에 남은 '('가 없는데, ')'가 입력된 경우
    public static boolean checkParenthesis(String parenthesis){
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < parenthesis.length(); i++) {
            char c = parenthesis.charAt(i);
            if(c=='('){
                stack.push(c);
            }else{
                if(stack.empty()) return false;
                char e = stack.pop();
                if (e == ')') return false;
            }
        }

        // 모든 문자열을 처리했는데, stack 안에 원소가 남아 있으면
        // 괄호 쌍이 안 맞는 것
        if(!stack.empty()) return false;

        return true;
    }
}
